src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/PopulateLicenceDeletedAtMigration.kt

```diff
@@ -0,0 +1,106 @@
+package uk.gov.justice.digital.hmpps.hmppshdcapi.licences
+
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.data.domain.Limit
+import org.springframework.data.domain.Page
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.hmppshdcapi.licences.prison.Booking
+import uk.gov.justice.digital.hmpps.hmppshdcapi.licences.prison.PrisonApiClient
+import java.time.LocalDateTime
+
+val UNKNOWN_DELETED_AT = null
+
+@Service
+class PopulateLicenceDeletedAtMigration(
+  private val licenceRepository: LicenceRepository,
+  private val licenceVersionRepository: LicenceVersionRepository,
+  private val prisonApiClient: PrisonApiClient,
+) {
+
+  companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  @Transactional
+  fun run(previousLastIdProcessed: Long, numberToMigrate: Int = 1000) {
+    val licencesRecords = licencesToMigrate(previousLastIdProcessed, numberToMigrate)
+    do {
+      val lastIdProcessed = licencesRecords.content.last().first.id
+      log.info("Last Id processed in batch: $lastIdProcessed")
+      applyAnySoftDeletes(licencesRecords, numberToMigrate)
+    } while (!licencesRecords.isEmpty)
+  }
+
+  private fun licencesToMigrate(previousLastIdProcessed: Long, numberToMigrate: Int): Page<Pair<Licence, Booking?>> {
+    val hdcLicences = licenceRepository.findAllByDeletedAtAndIdGreaterThanLastProcessedAndOrderByIdAsc(UNKNOWN_DELETED_AT, previousLastIdProcessed, Limit.of(numberToMigrate))
+    val bookings = getBookings(hdcLicences)
+    return hdcLicences.map { it to bookings[it.bookingId] }
+  }
+
+  private fun getBookings(hdcLicences: Page<Licence>): Map<Long, Booking> {
+    val bookings = hdcLicences.content.mapNotNull { prisonApiClient.getBooking(it.bookingId) }
+    return bookings.associateBy { it.bookingId }
+  }
+
+  fun isToBeSoftDeleted(booking: Booking?, today: LocalDateTime): Boolean {
+    val topupSupervisionExpiryDate = booking?.topupSupervisionExpiryDate
+    val licenceExpiryDate = booking?.licenceExpiryDate
+    if (topupSupervisionExpiryDate != null && licenceExpiryDate != null) {
+      if (topupSupervisionExpiryDate < licenceExpiryDate) {
+        val isLEDTodayOrPast = licenceExpiryDate <= today
+        if (isLEDTodayOrPast) return true
+      }
+    }
+    if (topupSupervisionExpiryDate != null) {
```

Comment:
if `!isLEDTodayOrPast` and both are present then we will still check this rule.. is that correct or should we continue to try the next rule? 

At the moment the whole thing is: 
```kt
return 
     (topupSupervisionExpiryDate < licenceExpiryDate &&  licenceExpiryDate <= today) ||
     (topupSupervisionExpiryDate <= today) ||
     (licenceExpiryDate <= today)
```

which could be rewritten as: 
```kt
   return licenceExpiryDate <= today || topupSupervisionExpiryDate <= today
```

Reply:
or should you be returning from each case: 
```kt
if (topupSupervisionExpiryDate != null && licenceExpiryDate != null) {
      if (topupSupervisionExpiryDate < licenceExpiryDate) {
        return licenceExpiryDate <= today
     }
}
```

Reply:
I've been thinking about this. Shirley's rules are: 

Rules:
At midnight check:

IF no TUSED
Then soft delete when LED = today OR is in the past

IF TUSED present
Then soft delete when TUSED = today OR is in the past
* IF TUSED is before LED
* Then use LED

Licence end date is:
* TUSED if present
* else LED

Which I've interpreted as:

IF no TUSED
Then soft delete when LED = today OR is in the past

IF TUSED present but TUSED is before LED
Then soft delete when LED = today OR is in the past

IF TUSED present
Then soft delete when TUSED = today OR is in the past

I think we should be returning from each case. Done a little refactoring as suggested. Hope this is correct.